File: internal/core/application/usecases/update_payment_status.go
package usecases

import (
	"context"
	"fmt"

	"github.com/fiap/ms-payment/internal/core/domain/entities"
	"github.com/fiap/ms-payment/internal/core/domain/ports"
	"github.com/fiap/ms-payment/pkg/utils"
)

type UpdatePaymentStatusUseCase struct {
	paymentRepository        ports.PaymentRepository
	generateTokenMP          ports.MercadoPagoGenerateToken
	getPaymentMP             ports.MercadoPagoGetPayment
	updatePaymentStatusTopic ports.PaymentStatusTopic
}

func NewUpdatePaymentStatusWebhookUseCase(
	paymentRepository ports.PaymentRepository,
	generateTokenMP ports.MercadoPagoGenerateToken,
	getPaymentMP ports.MercadoPagoGetPayment,
	updatePaymentStatusTopic ports.PaymentStatusTopic,
) *UpdatePaymentStatusUseCase {
	return &UpdatePaymentStatusUseCase{
		paymentRepository:        paymentRepository,
		generateTokenMP:          generateTokenMP,
		getPaymentMP:             getPaymentMP,
		updatePaymentStatusTopic: updatePaymentStatusTopic,
	}
}

func (u *UpdatePaymentStatusUseCase) Execute(ctx context.Context, paymentId string, externalId string) error {
	payment, err := u.paymentRepository.GetByExternalId(ctx, externalId)
	if err != nil {
		return utils.HTTPInternalServerError("failed to get payment by external ID")
	}

	if payment.Id == "" {
		fmt.Println("Payment not found")
		return nil
	}

	token, err := u.generateTokenMP.GenerateToken(ctx)
	if err != nil {
		return utils.HTTPPreconditionFailed("failed to generate token")
	}

	status, err := u.getPaymentMP.GetPaymentMock(ctx, paymentId, token)
	if err != nil {
		return utils.HTTPPreconditionFailed("failed to get payment")
	}

	mappedStatus := u.mapStatus(status)
	if payment.Status == mappedStatus {
		fmt.Println("Payment status already up to date")
		return nil
	}

	err = u.paymentRepository.UpdateStatusByExternalId(ctx, externalId, string(mappedStatus))
	if err != nil {
		return utils.HTTPInternalServerError("failed to update payment status")
	}

	err = u.updatePaymentStatusTopic.PublishPaymentStatus(ctx, externalId, string(mappedStatus))
	if err != nil {
		return utils.HTTPInternalServerError("failed to publish payment status")
	}

	return nil
}

func (u *UpdatePaymentStatusUseCase) mapStatus(status string) entities.PaymentStatus {
	switch status {
	case "pending", "authorized", "in_process":
		return entities.PaymentStatus_Pending
	case "approved", "in_mediation":
		return entities.PaymentStatus_Paid
	case "cancelled", "refunded", "charged_back":
		return entities.PaymentStatus_Canceled
	default:
		return entities.PaymentStatus_Rejected
	}
}

File: src/core/application/usecases/update_payment_status.ts
import { IMercadoPagoGenerateToken } from "@ports/generate_token_mp"
import { HTTPNotFound, HTTPPreconditionFailed } from "@utils/http"
import { IPaymentRepository } from "@ports/payment_repository"
import { IMercadoPagoGetPayment } from "@ports/get_payment_mp"
import { PaymentStatus } from "@entities/payment.types"
import { IPaymentTopic } from "@ports/payment_topic"

export class UpdatePaymentStatusUseCase {
	private readonly paymentRepository: IPaymentRepository
	private readonly generateTokenMP: IMercadoPagoGenerateToken
	private readonly getPaymentMP: IMercadoPagoGetPayment
	private readonly paymentTopic: IPaymentTopic

	constructor(
		paymentRepository: IPaymentRepository,
		generateTokenMP: IMercadoPagoGenerateToken,
		getPaymentMP: IMercadoPagoGetPayment,
		paymentTopic: IPaymentTopic,
	) {
		this.paymentRepository = paymentRepository
		this.generateTokenMP = generateTokenMP
		this.getPaymentMP = getPaymentMP
		this.paymentTopic = paymentTopic
	}

	async execute(paymentId: string, externalId: string): Promise<void> {
		const payment = await this.paymentRepository.getByExternalId(externalId)
		if (!payment) {
			throw new HTTPNotFound("Payment not found")
		}

		const token = await this.generateTokenMP.generateToken()
		if (!token) {
			throw new HTTPPreconditionFailed("Failed to generate token")
		}

		const status = await this.getPaymentMP.getPaymentStatusMock(paymentId, token)
		if (!status) {
			throw new HTTPPreconditionFailed("Failed to get payment")
		}

		const mappedStatus = this.mapStatus(status)
		if (payment.status === mappedStatus) {
			console.info("Payment status already up to date")
			return
		}

		await this.paymentRepository.updateStatusByExternalId(externalId, mappedStatus)
		await this.paymentTopic.publishPaymentStatus(externalId, mappedStatus)
	}

	private mapStatus(status: string): PaymentStatus {
		switch (status) {
			case "pending":
			case "authorized":
			case "in_process":
				return PaymentStatus.Pending
			case "approved":
			case "in_mediation":
				return PaymentStatus.Paid
			case "cancelled":
			case "refunded":
			case "charged_back":
				return PaymentStatus.Canceled
			default:
				return PaymentStatus.Rejected
		}
	}
}